python_modules/dagster/dagster/core/types/field.py

```diff
@@ -0,0 +1,209 @@
+
+        self.description = check.opt_str_param(description, 'description')
+        if is_optional == INFER_OPTIONAL_COMPOSITE_FIELD:
+            is_optional = all_optional_type(self.config_type)
+            if is_optional is True:
+                from .evaluator import hard_create_config_value
+
+                self._default_value = lambda: hard_create_config_value(self.config_type, None)
+            else:
+                self._default_value = default_value
+        else:
+            is_optional = check.bool_param(is_optional, 'is_optional')
+            self._default_value = default_value
+
+        if is_optional is False:
+            check.param_invariant(
+                default_value == FIELD_NO_DEFAULT_PROVIDED,
+                'default_value',
+                'required arguments should not specify default values',
+            )
+
+        self.is_optional = is_optional
+
+    @property
+    def is_required(self):
+        return not self.is_optional
+
+    @property
+    def default_provided(self):
+        '''Was a default value provided
+
+        Returns:
+            bool: Yes or no
+        '''
+        return self._default_value != FIELD_NO_DEFAULT_PROVIDED
+
+    @property
+    def default_value(self):
+        check.invariant(self.default_provided, 'Asking for default value when none was provided')
+
+        if callable(self._default_value):
+            return self._default_value()
+
+        return self._default_value
+
+    @property
+    def default_value_as_str(self):
+        check.invariant(self.default_provided, 'Asking for default value when none was provided')
+
+        if callable(self._default_value):
+            return repr(self._default_value)
+
+        return str(self._default_value)
+
+
+def get_config_cls(builtin_enum):
+    check.inst_param(builtin_enum, 'builtin_enum', BuiltinEnum)
+    return _CONFIG_CLS_MAP[builtin_enum]
+
+
+_CONFIG_CLS_MAP = {
+    BuiltinEnum.ANY: Any,
+    BuiltinEnum.STRING: String,
+    BuiltinEnum.INT: Int,
+    BuiltinEnum.BOOL: Bool,
+    BuiltinEnum.PATH: Path,
+}
+
+
+class _ConfigHasFields(ConfigType):
+    def __init__(self, fields, *args, **kwargs):
+
+        self.fields = check.dict_param(fields, 'fields', key_type=str, value_type=Field)
+        super(_ConfigHasFields, self).__init__(*args, **kwargs)
+
+    @property
+    def inner_types(self):
+        return list(set(self._yield_inner_types()))
+
+    def _yield_inner_types(self):
+        for field in self.fields.values():
+            yield field.config_type
+            for inner_type in field.config_type.inner_types:
+                yield inner_type
+
+
+class ConfigComposite(_ConfigHasFields):
+    @property
+    def is_composite(self):
+        return True
+
+
+class ConfigSelector(_ConfigHasFields):
+    @property
+    def is_selector(self):
+        return True
+
+
+# HACK HACK HACK
+#
+# This is not good and a better solution needs to be found. In order
```

Comment:
This is having consequences for error messaging in dagit as well -- we saw this morning during a demo where `Dict.4` (a meaningless name for the user) was implicated in an otherwise helpful message in the config editor.

Reply:
https://github.com/dagster-io/dagster/issues/575